src/rust/engine/src/scheduler.rs

```diff
@@ -26,33 +26,75 @@ use ui::EngineDisplay;
 /// Both Scheduler and Session are exposed to python and expected to be used by multiple threads, so
 /// they use internal mutability in order to avoid exposing locks to callers.
 ///
-pub struct Session {
+struct InnerSession {
   // The total size of the graph at Session-creation time.
   preceding_graph_size: usize,
   // The set of roots that have been requested within this session.
   roots: Mutex<HashSet<Root>>,
   // If enabled, the display that will render the progress of the V2 engine.
   display: Option<Mutex<EngineDisplay>>,
+  // If enabled, Zipkin spans for v2 engine will be collected.
+  should_record_zipkin_spans: bool,
+  // A place to store info about workunits in rust part
+  workunits: Mutex<Vec<WorkUnit>>,
+}
+
+#[derive(Clone)]
+pub struct Session(Arc<InnerSession>);
```

Comment:
It's a fairly cosmetic thing - there's a convention to make an outer/inner thing to make something cloneable (particularly if your type needs to implement traits, because you can't implement traits on `Arc<_>` unless the trait happens to be defined in your crate), but also passing around `Arc`s to things is generally fine :)

Reply:
I also just find the use of `.0` to be difficult to follow personally, and thought that having an `Arc<Session>` instead might make this diff smaller and reduce the number of passthrough helper methods. The code is quite readable either way, and the current diff will make it easier to introduce some more methods on `Session` later.

Reply:
I still don't know that I see the utility of an `InnerSession` struct when it requires so many extra accessors such as `preceding_graph_size`, compared to just using an `Arc<Session>`. It seems to inflate the diff count and make it more difficult to symbol search. This PR can be merged regardless, I just don't quite understand why it was architected this way.

Reply:
The main idea was to point out that Session now is Arc<Sesion> and it is so because of workunits. In this case, I explicitly say in the definition of the Session that it is Arc (not when it is used as a part of another struct). The Session is used in RootContext and Context (at least two places) and in both places, I need to put Arc<Session>. It is not difficult, but not obvious why the Session should be Arc<Session> in these places.